### PR TITLE
Param builder better honours securely allocated memory

### PR DESCRIPTION
**Pull request by paulidale**
Add additional checks for the string types being in secure memory and respect this if so.

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by mattcaswell — APPROVED** [review]

File: test/param_build_test.c [hunk 1]
```diff
@@ -126,8 +128,14 @@ static int template_private_test(void)
...
-    if (!TEST_ptr(bld)
-        || !TEST_true(OSSL_PARAM_BLD_push_uint(bld, "i", 6))
+    if (!TEST_ptr(data1 = OPENSSL_secure_malloc(data1_size))
+            || !TEST_ptr(bld = OSSL_PARAM_BLD_new()))
+        goto err;
+
+    for (j = 0; j < data1_num; j++)
+        data1[j] = -16 * j;
+
+    if (!TEST_true(OSSL_PARAM_BLD_push_uint(bld, "i", 6))
```
Comment (t8m) [inline test/param_build_test.c:138]: Should there be some CRYPTO_secure_allocated() calls to test that the relevant params `i` and `bignumber` are securely allocated?

Comment (paulidale) [reply test/param_build_test.c:138]: The native types are function arguments passed by value, so they can't possibly be in secure memory.

The BNs are checked via the `BN_FLG_SECURE` flag.  The BN structure shouldn't ever be in secure memory but the number bytes themselves can be.

Pauli

Comment (paulidale) [reply test/param_build_test.c:138]: BN's already supported secure allocation in param_bld.

Comment (t8m) [reply test/param_build_test.c:138]: I am sorry for a confusion - I meant `oct_s` instead of `i`.

Comment (t8m) [reply test/param_build_test.c:138]: Also there is no check that the bn_res has BN_FLG_SECURE or am I missing something?

Comment (paulidale) [reply test/param_build_test.c:138]: The secure memory allocation code is checked by `test_secmem`, is it necessary to verify it here too?

Comment (t8m) [reply test/param_build_test.c:138]: That tests the low level secure memory allocation. What I am asking here for is to test that the OSSL_PARAM_BLD honors the secure memory allocation of the origin.

Comment (paulidale) [reply test/param_build_test.c:138]: Okay, I understand finally.  I will add.

Comment (paulidale) [reply test/param_build_test.c:142]: Done.

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master, thanks for the reviews.

**Pull request closed without merging**